src/Typechecker/Typechecker.hs

```diff
@@ -105,23 +105,15 @@ exprtype (Ref s) = do
 exprtype (Tuple xs) = do
   xs' <- mapM exprtype xs
   return $ Tup xs'
-exprtype (App n es) = do -- FIXME. Tuple composition is bad.
-  es' <- exprtype es
-  let es'' = case es' of
-        Tup [Tup xs] -> Tup xs
-        x -> x
```

Comment:
Okay, so I see how this helps out now. Lots of tuple repack/unpack, and now we can obviate that behavior, nice!